src/util/hass-translation.ts

```diff
@@ -1,58 +1,89 @@
 import { translationMetadata } from "../resources/translations-metadata";
+import { fetchFrontendUserData } from "../data/frontend";
+import { HomeAssistant } from "../types";
+
+const STORAGE = window.localStorage || {};
+
+// Chinese locales need map to Simplified or Traditional Chinese
+const LOCALE_LOOKUP = {
+  "zh-cn": "zh-Hans",
+  "zh-sg": "zh-Hans",
+  "zh-my": "zh-Hans",
+  "zh-tw": "zh-Hant",
+  "zh-hk": "zh-Hant",
+  "zh-mo": "zh-Hant",
+  zh: "zh-Hant", // all other Chinese locales map to Traditional Chinese
+};
+
+/**
+ * Search for a matching translation from most specific to general
+ */
+function findAvailableLanguage(language: string) {
+  // In most case, the language has the same format with our translation meta data
+  if (language in translationMetadata.translations) {
+    return language;
+  }
 
-export function getActiveTranslation() {
   // Perform case-insenstive comparison since browser isn't required to
   // report languages with specific cases.
-  const lookup = {};
-  /* eslint-disable no-undef */
-  Object.keys(translationMetadata.translations).forEach((tr) => {
-    lookup[tr.toLowerCase()] = tr;
-  });
+  const langLower = language.toLowerCase();
 
-  // Search for a matching translation from most specific to general
-  function languageGetTranslation(language) {
-    const lang = language.toLowerCase();
+  if (langLower in LOCALE_LOOKUP) {
+    return LOCALE_LOOKUP[langLower];
+  }
 
-    if (lookup[lang]) {
-      return lookup[lang];
+  for (const lang in Object.keys(translationMetadata.translations)) {
+    if (lang.toLowerCase() === langLower) {
+      return lang;
     }
-    if (lang.split("-")[0] === "zh") {
-      return lang === "zh-cn" || lang === "zh-sg" ? "zh-Hans" : "zh-Hant";
+  }
+}
+
+/**
+ * Get user selected language from backend
+ */
+export async function getUserLanguage(hass: HomeAssistant) {
+  const userLanguage = await fetchFrontendUserData(hass, "language");
+  if (userLanguage) {
+    const language = findAvailableLanguage(userLanguage);
+    if (language) {
+      return language;
     }
-    return null;
   }
+  return getLocalLanguage();
```

Comment:
We should not do this. This means that the calling code cannot differentiate between a chosen and autodetected language. It means that we will, on initial load, store the local language as the users preference, instead of sticking to the default detected language.

Reply:
it would not save as user preference, since we never save it back unless user selected language through UI. But I changed it to `return null` anyway, I agreed that `null` is more clearly.